File: C study/make the Dduck.cpp
#include <bits/stdc++.h>

using namespace std;

//떡의 개수(N)과 요청한 떡의 길이(M)
int n, m;

// 각 떡의 개별 높이 정보
vector<int> arr;

int main(void){
	cin>>n>>m;
	for (int i=0;i<n;i++){
		int x;
		cin>>x;
		arr.push_back(x);
	}
	//이진 탐색을 위한 시작점과 끝점 설정
	int start=0;
	int end=1e9;
	// 이진 탐색 수행을 반복적으로 시행한다.
	int result=0;
	while(start<=end){
		long long int total=0;
		int mid=(start+end)/2;
		for(int i=0;i<n;i++){
			//잘랐을 때의 떡의 양 계산
			if(arr[i]>mid) total+=arr[i]-mid; 
		}
		if(total<m){ //떡의 양이 만약 부족하다면 더 많이 잘라야 하므로 왼쪽 부분 탐색한다.
			end=mid-1; 
		}
		else{
			//떡의 양이 충분하다면 덜 잘라야 하므로 오른쪽 부분을 탐색한다.
			result=mid;
			start=mid+1; 
		}
	} 
	cout<<result<<'\n';
} 

File: Python/2021년/Make the DDuck.py
# 떡의 개수(N)와 요청한 떡의 길이(M)을 입력받는다.
n,m=list(map(int,input().split()))
# 각 떡의 개별 높이 정보를 입력받는다.
array=list(map(int,input().split()))

# 이진 탐색을 위한 시작점과 끝점을 설정한다.
start=0
end=max(array)

# 이진 탐색을 반복적으로 수행한다.
result=0
while(start<=end):
    total=0
    mid=(start+end)//2
    for x in array:
        # 잘랐을 때의 떡의 양을 계산한다.
        if x>mid:
            total+=x-mid
    # 떡의 양이 부족한 경우에는 더 많이 자른다.(왼쪽 부분을 탐색)
    if total<m:
        end=mid-1
    # 떡의 양이 충분한 경우에는 덜 자른다.(오른쪽 부분을 탐색)
    else:
        result=mid # 최대한 덜 잘랐을 때가 정답이므로, 여기에서 result에 기록한다.
        start=mid+1

print(result)